plugins/inputs/ipvs/ipvs.go

```diff
@@ -0,0 +1,114 @@
+// +build linux
+
+package ipvs
+
+import (
+	"errors"
+	"fmt"
+	"math/bits"
+	"strconv"
+	"syscall"
+
+	"github.com/docker/libnetwork/ipvs"
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/plugins/inputs"
+)
+
+// IPVS holds the state for this input plugin
+type IPVS struct {
+	handle *ipvs.Handle
+}
+
+// Description returns a description string
+func (i *IPVS) Description() string {
+	return "Collect virtual and real server stats from Linux IPVS"
+}
+
+// SampleConfig returns a sample configuration for this input plugin
+func (i *IPVS) SampleConfig() string {
+	return ``
+}
+
+// Gather gathers the stats
+func (i *IPVS) Gather(acc telegraf.Accumulator) error {
+
+	if i.handle == nil {
+		h, err := ipvs.New("") // TODO: make the namespace configurable
+		if err != nil {
+			return errors.New("Unable to open IPVS handle")
+		}
+		i.handle = h
+	}
+
+	services, err := i.handle.GetServices()
+	if err != nil {
+		i.handle.Close()
+		i.handle = nil // trigger a reopen on next call to gather
+		return errors.New("Failed to list IPVS services")
+	}
+	for _, s := range services {
+		fields := map[string]interface{}{
+			"connections": s.Stats.Connections,
+			"pkts_in":     s.Stats.PacketsIn,
+			"pkts_out":    s.Stats.PacketsOut,
+			"bytes_in":    s.Stats.BytesIn,
+			"bytes_out":   s.Stats.BytesOut,
+			"pps_in":      s.Stats.PPSIn,
+			"pps_out":     s.Stats.PPSOut,
+			"cps":         s.Stats.CPS,
+		}
+		acc.AddGauge("ipvs_virtual_server", fields, serviceTags(s))
+	}
+
+	return nil
+}
+
+// helper: given a Service, return tags that identify it
+func serviceTags(s *ipvs.Service) map[string]string {
+	ret := map[string]string{
```

Comment:
Add a tag `source` and set it to `os.Hostname()`, this probably seems somewhat pointless, but it is for #4413 and will be nice when we get all plugins supporting it.

Reply:
Since these are not metrics from a remote service (as described in #4413 ), I dont think this plugin should emit that tag.

As I understand that issue - the problem arises when metrics from a remote host (not the same host that telegraf is running on) are received and then forwarded by telegraf, we want to have a way to know what `data source` the metrics originated on.

However in this case, the metrics are from the host that telegraf is running on (like, cpu/mem/network). So the `agent.omit_hostname` (and related) config knobs should already provide this.

Please correct me if I'm missing something.

Reply:
In this case the `source` and `host` tag have the same value, but still would have different meanings.  Adding it in this case would just be for uniformity with other plugins with the goal that eventually all metrics would contain both of these tags.

With both tags you will be able to query either by the Telegraf system that reported the metric or by the system that the metric is about, without needing to check which is supported for each plugin, and will be able to list all metrics for a hostname more easily.

Reply:
@danielnelson  - I think I understand.

I think this conversation is probably to be had on #4413, but I'll continue here for now...

I'd like to point out that the nice thing about the `host` tag is/was that it defaulted to `os.Hostname` but was overridable via the config file. This thoughtful knob was immensely useful in situations where we deploy telegraf to large deployments where the logical name is better understood than the actual dns/hostname given to the system.

After reading the conversation on #4413 (which wandered in the direction of selecting an appropriate name) I still feel like the `source` tag should be added by telegraf when one is not provided by the input plugin (using whatever value it uses for the `host` tag). This makes a lot of sense for all the metrics that originate on the host itself and frees the plugin author from needing to think about this. 

I'd argue that the majority of the metrics collected by the input plugins are from the local host itself. There are fewer plugins that collect metrics from over the network (statsd, wavefront, etc).

If we proceed in this direction, we potentially run into a situation where every input plugin can now handle this tag with subtle differences. In one case, as a plugin author, I may choose to accept this as a config element so as to allow users to specify an alternate value, in other cases the user may override this using `[[inputs.foobar.tags]]`, thus making this redundant. In other cases, Ithe burden to add a `tagdrop` now falls on the user (where she doesn't want this tag). We're now beginning to fray one of telegrafs greatest strengths - its simplicity.

I'd hope that the burden of doing this falls only on the input plugins who regularly collect metrics from over the network (statsd, wavefront, etc).

Has a final decision been reached on this? If so, then I'll go ahead and add this as you suggest. If not, I'd like to participate in that discussion before I add this.

Reply:
There hasn't been any discussion about automatically adding the tag, so I would comment on 4413 about the idea and any details.  I'm not opposed to the idea, but it seems that if we automatically add the tag we need to go through all the remote plugins in one pass and add in the right value.   Otherwise on top of it being incorrect I think it will cause confusion about what the tag means.  I'd say it's about a 50/50 split between localhost only and plugins that can support a remote host on the input plugins.

For plugins that can collect remotely it will have to be the plugin who sets the value, as we don't have a uniform way to determine it from looking at the config and sometimes hostnames are dynamic.  Uniformity in handling will just have to be enforced manually for this reason, and that means we can't totally solve the issue by adding it automatically.

Filtering it out is pretty easy, just add a tagexclude on the outputs, but I think this is something that very few will want.  In general adding a tag like this doesn't increase series cardinality because it either doesn't vary with respect to the the host or there already is identifying information on the metric that it doesn't vary against.

We aren't very far along on adding this tag, so we can leave it out of this plugin for now.  Once we have proper documentation in place for how it should work then we will begin requiring it in new plugins.